drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+    BOOLEAN ListenCancelled;
+    HANDLE ListenThreadHandle;     /* needed to restart listening */
+    PKTHREAD ListenThread;
+    KEVENT StartListenEvent;
+    BOOLEAN ListenThreadShouldRun;
+
+    /* AcceptSocket's keep a reference on their listen sockets so
+     * that the Address file will be closed only if there are no
+     * open connections any more.
+     */
+    struct _WSK_SOCKET_INTERNAL *ListenSocket;
+
+    /* SocketPut() only works at PASSIVE_LEVEL. If > PASSIVE_LEVEL,
+     * it will put the socket on this list and wake the putsockets
+     * thread.
+     */
+    struct _WSK_SOCKET_INTERNAL *NextSocketToPut;
+    int NumSocketPuts;
+} WSK_SOCKET_INTERNAL, *PWSK_SOCKET_INTERNAL;
+
+struct NetioContext
+{
+    PIRP UserIrp;
+    PWSK_SOCKET_INTERNAL socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    PTDI_CONNECTION_INFORMATION PeerAddrRet;
+};
+
+void
+SocketGet(PWSK_SOCKET_INTERNAL s)
+{
+    FUNCTION_TRACE;
+
+    s->RefCount++;
+// DbgPrint("SocketGet: refcount is %d socket is %p\n", s->RefCount, s);
+}
+
+void
+SocketPut(PWSK_SOCKET_INTERNAL s);
+
+static
+void SocketShutdown(PWSK_SOCKET_INTERNAL s)
+{
+    NTSTATUS status;
+
+    FUNCTION_TRACE;
+
+    if (s->ListenSocket != NULL)
+    {
+        SocketPut(s->ListenSocket);
+        s->ListenSocket = NULL;
+    }
+    if (s->ListenThreadHandle != NULL)
+    {
+        s->ListenThreadShouldRun = FALSE;
+        KeSetEvent(&s->StartListenEvent, IO_NO_INCREMENT, FALSE);
+        status = KeWaitForSingleObject(s->ListenThread, Executive, KernelMode, FALSE, (PLARGE_INTEGER)NULL);
+        if (!NT_SUCCESS(status))
+        {
+            DbgPrint("KeWaitForSingleObject failed with status 0x%08x!\n", status);
+        }
+        ObDereferenceObject(s->ListenThread);
+        s->ListenThread = NULL;
+        s->ListenThreadHandle = NULL;
+    }
+    if (s->ListenIrp != NULL)
+    {
+        s->ListenCancelled = TRUE;
+        IoCancelIrp(s->ListenIrp);
+        s->ListenIrp = NULL;
+    }
+    if (s->ConnectionFile != NULL) {
+        if (s->ConnectionFileAssociated) {
+               /* This fails with error 0xc000023b (still connected) on Windows 2003 */
+            status = TdiDisassociateAddressFile(s->ConnectionFile);
+            if (!NT_SUCCESS(status)) {
+                NETIO_DbgPrint(MIN_TRACE, ("Warning: TdiDisassociateAddressFile returned status %08x\n", status));
+            }
+            s->ConnectionFileAssociated = FALSE;
+        }
+        ObDereferenceObject(s->ConnectionFile);
+        s->ConnectionFile = NULL;
+    }
+    if (s->ConnectionHandle != NULL)
+    {
+        ZwClose(s->ConnectionHandle);
+        s->ConnectionHandle = NULL;
+    }
+}
+
+static PWSK_SOCKET_INTERNAL SocketsToPut = NULL;
+static KSPIN_LOCK SocketsToPutListLock = 0;
+static BOOLEAN PutSocketsThreadShouldRun = FALSE;
+static KEVENT PutSocketsEvent;
+static HANDLE PutSocketsThreadHandle;
+
+static void WSKAPI PutSocketsThread(void *p)
```

Comment:
New driver code: please refer to point 6 of https://reactos.org/wiki/Coding_Style#Indentation_and_line_width .
(In particular, please use [SAL2 annotations](https://www.codeproject.com/Reference/879527/SAL-Function-Parameters-Annotations)